File: src/ruby/maximum_subarray.rb
def max_sub_array(nums)
  max = 0
  subarray = nums
  sum = 0

  neg_idx = subarray.find_index{|num| num < 0 }
  until neg_idx.nil?
    sum += subarray[0...neg_idx].sum
    max = sum if sum > max
    sum += subarray[neg_idx]
    sum = 0 if sum < 0
    subarray = subarray[(neg_idx+1)..-1]
    neg_idx = subarray.find_index{|num| num < 0 }
  end
  sum += subarray.sum
  [sum, max].max
end

puts max_sub_array([-2,1,-3,4,-1,2])
puts max_sub_array([-2,1,-3,4,-1,-1,2,1,-4,-2,6])

File: src/python/src/maximum_subarray.py
# https://leetcode.com/problems/maximum-subarray/description/?envType=problem-list-v2&envId=a6ezkna5&difficulty=MEDIUM%2CHARD


class Solution:
    def maxSubArray(self, nums: list[int]) -> int:
        max_sum = current_sum = nums[0]
        for num in nums[1:]:
            current_sum = max(current_sum + num, num)
            max_sum = max(max_sum, current_sum)
        return max_sum
